linker: warn about orphan sections

dcpleung opened the pull request:
This adds a linker flag and necessary changes to linker scripts so that linker will warn about orphan sections.

Relates to #5534

dcpleung commented:
Sanitycheck passed. But marked as DNM for now so we can gauge the impact.

andyross reviewed (approved):
So... devil's advocacy: that is a LOT of change just to enable one warning.  And it's likely this is going to have to evolve along with us as e.g. the toolchain grows new hair (random thought: does LTO involve a bunch of new section names?  I bet it does).

I mean, it's a good warning.  But we're biting off a bunch of linker complexity to enable it.  Not sure the tradeoff is worth it personally.  But then again, you've already done all the work :)

dcpleung commented:
There is an issue #9498 where the developer specifies a custom section which results in fault. The reason was because the linker "randomly" places the custom section in front of data/bss section at final linking. This moves the device structs around which invalidates the addresses being calculated using the prebuilt binary So the idea is to warn at build time (#5534). A lot of these sections (like the debug.* and rel.*) are included in the linker scripts by binutils to produce ELF binaries. And yes, new sections produced by compilers will now need to be specified manually.

nashif commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -226,6 +270,19 @@ SECTIONS
 	} GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
 #endif
 
+#ifdef CONFIG_SOC_SERIES_PSOC62
```
looks like we need a better way to add soc specific sections into this general .ld file. Adding SoC specific section will make this file explode with ifdefs.

dcpleung replied:
Added something similar to the custom-*.ld for SoC.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -287,6 +287,13 @@ zephyr_ld_options(
 
+# Funny thing is if you set this to =error, some architectures will
+# fail the compiler flag check and this will be removed from build.
+# So keep it at =warn.
```
Would

zephyr_link_libraries(
  ${LINKERFLAGPREFIX},--orphan-handling=error
  )

have the behaviour we want?

dcpleung replied:
The linker will fail if =error is used when there are orphan sections. However, the ARC and Xtensa, cmake somehow silently drops the flag. There is nothing to indicate problems in the CMakeError.log. Deliberately setting the wrong parameter (e.g. =badarg) would result in error output about invalid argument in CMakeError.log. I haven't been able to figure out why cmake would drop the flag. So we can only use =warn for now to enable warning for all architectures. Sanitycheck will still errors out if there are any orphan sections.

SebastianBoe replied:
I was able to reproduce your observation about the flag being dropped with zephyr_ld_options, but when using zephyr_link_libraries I got the desired behaviour. Error on orphan sections.

Note, using zephyr_link_libraries should be used in place of zephyr_ld_options.

zephyr_ld_options will by-design drop the flag if cmake is unable to compile a simple C file with it (presumably there are some orphan sections in the compiler flag test).

zephyr_link_libraries just adds the flag without any toolchain-capability test.

https://gist.github.com/SebastianBoe/d480b62d0f998d0f108ad75fba8ed1fc

zephyr-sdk-0.9.3
qemu_xtensa

dcpleung replied:
I thought about it but I am hesitant on forcing this flag without compiler check. Putting this in zephyr_link_libraries assumes everyone will be using binutils to link the final binaries. As we support custom toolchain, that compiler/linker combo may not support this flag.

SebastianBoe reviewed (approved):
This improves the error feedback significantly for non-advanced users at the cost of added maintenance for advanced users, so this is very welcome.

SebastianBoe commented on soc/Kconfig:
```diff
@@ -11,3 +11,24 @@ osource "$(SOC_DIR)/$(ARCH)/Kconfig"
+config SOC_NOINIT_LD
+	bool
+	help
+	  Include a SoC specific linker script fragment for inserting
```
a SoC specific linker script fragment

an SoC specific linker script fragment named soc-noinit.ld

SebastianBoe commented on soc/Kconfig:
```diff
@@ -11,3 +11,24 @@ osource "$(SOC_DIR)/$(ARCH)/Kconfig"
 endmenu
+
+#
+# SOC_*_LD: SoC specific Linker script additions
```
There is no reasonable way for a non-ARM user to discover that he can't use this feature. We either need to make the Kconfig options depend on ARM or add support for all platforms.

dcpleung replied:
Added the #include for common linker scripts under include/, and added conditionals for the Kconfig.

dcpleung committed:
arm: cypress/psoc6: add SoC specific linker input sections

The Cypress PSoC6 specifies some input sections in the startup
scripts. These sections (.heap, .stack, etc.) need to be placed
at correct location.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: warn about orphan sections

This adds a linker flag and necessary changes to linker scripts
so that linker will warn about orphan sections.

Relates to #5534.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: allow SoC to insert linker script fragments

This allows the SoC to specify some additional linker script
fragments into the bss, data and read-only data sections.

For example, the Cypress PSOC6 has a few input sections that
must be put into bss and data sections. Without specifying
these in the linker script, they are consider orphan sections
and the placement is based on linker heuristic which is
arbitrary.

POSIX is not supported as the main linker script is
provided by the host system's binutils and we have no control
over it. Also, currently Xtensa SoCs have their own linker
scripts so there is no need to this feature.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: arm: add missing period for CCM section name

The CCM section name macros are missing the leading period.
Add the period so those items supposed to be in CCM sections
are actually being placed correctly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: wdt_basic_api: rename noinit section

The linker section where data is not being initialized is
.noinit.*.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

pfalcon commented:
@OlegKapshii: FYI, there were some psoc6 related changes in this PR (just so you know they were done).

SebastianBoe commented:
For future reference, this PR may cause regressions.

We may discover that some of the input sections that we are now explicitly placing in output sections were previously placed in a different and accidentally "correct" position, or should actually be discarded instead of placed.

pfalcon commented:
~~~
[100%] Linking C executable zephyr.elf
/usr/bin/ld: warning: orphan section `.rel._k_mutex.static.pthread_key_lock' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel._k_mem_slab.static.posix_timer_slab' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel._k_sem.static.pthread_key_sem' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel._k_sem.static.mq_sem' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel.log_const_kernel' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
Generating files from zephyr.elf for board: native_posix
~~~

native_posix. @aescolar: FYI

aescolar commented:
Thanks @pfalcon , but this would seem to be a bug in this PR that needs to be addressed by @dcpleung . @pfalcon could you create an issue an assign it to him?

pfalcon commented:
> Thanks @pfalcon , but this would seem to be a bug in this PR

Well, maybe just misfeature, the resulting executables seems to work.

> @pfalcon could you create an issue an assign it to him?

Done: https://github.com/zephyrproject-rtos/zephyr/issues/10493
